driver-core/src/main/com/mongodb/internal/session/ServerSessionPool.java

```diff
@@ -111,14 +112,17 @@ private void closeSession(final ServerSessionImpl serverSession) {
             return;
         }
 
-        closedSessionIdentifiers.add(serverSession.getIdentifier());
+        synchronized (this) {
```

Comment:
Generally if you're protecting writes of a non-thread-safe data structure, you also have to protect reads.  That includes calls to size() and the implicit copy when constructing a new array from the shared array.

Try synchronizing everything except the call to `endClosedSessions` (this will require an extra local variable to hold the copied array).

Reply:
Done

Reply:
Consider the case where two threads call this method concurrently when the size of the array is 2 less than the max.  Both threads add to the array, then both threads check the size, and the size equals the max.  Then both threads will enter the conditional and end the same set of sessions